src/main/java/com/google/devtools/build/lib/syntax/Identifier.java

```diff
@@ -61,7 +62,9 @@ public void prettyPrint(Appendable buffer) throws IOException {
   public boolean equals(@Nullable Object object) {
     if (object instanceof Identifier) {
       Identifier that = (Identifier) object;
-      return this.name.equals(that.name);
+      // use identity check for improved performance as long as all names are canonicalized
+      // This is currently ensured by the static factory method creating Identifiers.
+      return this.name == that.name;
```

Comment:
Interning is entirely unnecessary for the correctness of the name resolution algorithm. Does this function show up in a profile? I would be astonished if so; why are identifiers ever compared for equality?

Reply:
because, since identifier names are now interned, their equality can now be checked using identity check instead of a value equality check

Reply:
But how does that help? If the answer is "performance", then that should be supported by evidence because (a) the change adds a cost, namely, the type of Identifier.name is no longer String but the more complex type "String interned by a particular interner", and (b) it is far from obvious that the overhead of interning these strings will be successfully amortized by subsequent cheaper equality tests.

Why does the performance of Identifier.equals matter anyway?

Reply:
@alandonovan, the reason why I've added this optimization was solely based on a comment from the original authors https://source.bazel.build/bazel/+/master:src/main/java/com/google/devtools/build/lib/syntax/Identifier.java;l=27?q=Identifier.java and is not the main purpose of this change, so as I've said above, I can separate this change so that it can be discussed separately.

Reply:
That seems best. The main point of the change, using Identifiers instead of Strings, is a good step.